package moriz.orangesunshine.fluid;

import moriz.orangesunshine.PSTags;
import moriz.orangesunshine.config.PSConfig;
import moriz.orangesunshine.entity.drug.DrugType;
import moriz.orangesunshine.entity.drug.influence.DrugInfluence;
import moriz.orangesunshine.fluid.alcohol.DrinkTypes;
import moriz.orangesunshine.fluid.alcohol.Maturity;
import moriz.orangesunshine.fluid.container.FluidContainer;
import moriz.orangesunshine.fluid.container.MutableFluidContainer;
import moriz.orangesunshine.fluid.container.Resovoir;
import moriz.orangesunshine.fluid.physical.FluidStateManager;
import moriz.orangesunshine.util.MathUtils;
import net.minecraft.ChatFormatting;
import net.minecraft.network.chat.Component;
import net.minecraft.resources.Identifier;
import net.minecraft.util.Mth;
import net.minecraft.world.item.Item;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import net.minecraft.world.level.block.state.properties.IntegerProperty;

import java.util.List;
import java.util.Objects;
import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

import org.jetbrains.annotations.Nullable;

import com.google.common.base.Suppliers;

/**
 * Created by lukas on 25.11.14.
 */
public class AlcoholicFluid extends DrugFluid implements Processable {
    public static final Attribute<Integer> DISTILLATION = Attribute.ofInt("distillation", 0, 16);
    public static final Attribute<Integer> MATURATION = Attribute.ofInt("maturation", 0, 16);

    private static final int FERMENTATION_STEPS = 2;
    public static final Attribute<Integer> FERMENTATION = Attribute.ofInt("fermentation", 0, FERMENTATION_STEPS);
    public static final Attribute<Boolean> VINEGAR = Attribute.ofBoolean("vinegar");

    final moriz.orangesunshine.fluid.AlcoholicFluid.Settings settings;

    public AlcoholicFluid(Identifier id, moriz.orangesunshine.fluid.AlcoholicFluid.Settings settings) {
        super(id, settings.drinkable().with(createVariantProperty(settings)));
        this.settings = settings;
    }

    private static FluidStateManager.FluidProperty<Integer> createVariantProperty(moriz.orangesunshine.fluid.AlcoholicFluid.Settings settings) {
        int maxVariant = Math.max(0, settings.states.get().size() - 1);
        return new FluidStateManager.FluidProperty<>(IntegerProperty.create("variant", 0, maxVariant), (stack, variant) ->
                settings.states.get().get(Mth.clamp(variant, 0, maxVariant)).apply(stack),
                stack -> settings.states.get().stream()
                        .filter(s -> s.entry().predicate().test(stack))
                        .findFirst()
                        .map(match -> settings.states.get().indexOf(match))
                        .orElse(0)
        );
    }

    protected int getDistilledColor() {
        return settings.distilledColor;
    }

    protected int getMatureColor() {
        return settings.matureColor;
    }

    @Override
    public void getDrugInfluencesPerLiter(ItemStack stack, Consumer<DrugInfluence> consumer) {
        super.getDrugInfluencesPerLiter(stack, consumer);

        double alcohol =
                  settings.fermentationAlcohol * (FERMENTATION.get(stack) / (double) FERMENTATION_STEPS)
                + settings.distillationAlcohol * MathUtils.progress(DISTILLATION.get(stack))
                + settings.maturationAlcohol * MathUtils.progress(MATURATION.get(stack) * 0.2F);

        consumer.accept(new DrugInfluence(settings.drugType, 20, 0.003, 0.002, alcohol));
        settings.variants.find(stack).extraDrug().ifPresent(drug -> {
            consumer.accept(drug.clone());
        });
    }

    @Override
    public int getProcessingTime(Resovoir tank, ProcessType type, @Nullable Resovoir complement) {
        if (type == ProcessType.DISTILL) {
            if (FERMENTATION.get(tank.getContents()) < FERMENTATION_STEPS || MATURATION.get(tank.getContents()) != 0) {
                return UNCONVERTABLE;
            }

            return settings.tickInfo.get().ticksPerDistillation;
        }

        if (type == ProcessType.MATURE) {
            if (FERMENTATION.get(tank.getContents()) < FERMENTATION_STEPS) {
                return UNCONVERTABLE;
            }
            return settings.tickInfo.get().ticksPerMaturation;
        }

        if (type == ProcessType.FERMENT) {
            if (FERMENTATION.get(tank.getContents()) < FERMENTATION_STEPS) {
                return settings.tickInfo.get().ticksPerFermentation;
            }
            return settings.tickInfo.get().ticksUntilAcetification;
        }

        return UNCONVERTABLE;
    }

    @Override
    public ItemStack process(Resovoir tank, ProcessType type, @Nullable Resovoir complement) {
        MutableFluidContainer contents = tank.getContents();

        if (type == ProcessType.DISTILL) {
            int fermentation = FERMENTATION.get(contents);


            if (fermentation < FERMENTATION_STEPS) {
                return ItemStack.EMPTY;
            }

            int distillation = DISTILLATION.get(contents);

            DISTILLATION.set(contents, distillation + 1);

            contents.drain(Mth.floor(contents.getLevel() * MathUtils.progress(distillation, 0.5F)));
            return PSFluids.SLURRY.getDefaultStack(1);
        }

        if (type == ProcessType.MATURE) {
            MATURATION.set(contents, MATURATION.get(contents) + 1);
        }

        if (type == ProcessType.FERMENT) {
            int fermentation = FERMENTATION.get(contents);

            if (fermentation < FERMENTATION_STEPS) {
                FERMENTATION.set(contents, fermentation + 1);
            } else {
                VINEGAR.set(contents, true);
            }
        }

        return ItemStack.EMPTY;
    }

    @Override
    public void getProcessStages(ProcessType type, ProcessStageConsumer consumer) {
        if (type == ProcessType.DISTILL) {
            generateRecipeConversions(settings.tickInfo.get().ticksPerDistillation, DISTILLATION,
                    DrinkTypes.State::distillation,
                    DrinkTypes.State::maturation,
                    DrinkTypes.State::fermentation, consumer);
        }

        if (type == ProcessType.MATURE) {
            generateRecipeConversions(settings.tickInfo.get().ticksPerMaturation, MATURATION,
                    DrinkTypes.State::maturation,
                    DrinkTypes.State::distillation,
                    DrinkTypes.State::fermentation, consumer);
        }

        if (type == ProcessType.FERMENT) {
            generateRecipeConversions(settings.tickInfo.get().ticksPerFermentation, FERMENTATION,
                    DrinkTypes.State::fermentation,
                    DrinkTypes.State::distillation,
                    DrinkTypes.State::maturation, consumer);
        }
    }

    private void generateRecipeConversions(int time, Attribute<Integer> attribute, Function<DrinkTypes.State, Integer> valueGetter,
            Function<DrinkTypes.State, Integer> fixA,
            Function<DrinkTypes.State, Integer> fixB,
            ProcessStageConsumer consumer) {
        List<DrinkTypes.State> states = settings.states.get();

        for (int i = 0; i < states.size(); i++) {
            var state = states.get(i);

            if (!state.vinegar()) {
                states.stream()
                        .filter(s -> {
                            return fixA.apply(s) == fixA.apply(state)
                                    && fixB.apply(s) == fixB.apply(state)
                                    && !s.vinegar();
                        })
                        .forEach(s -> {
                    int difference = valueGetter.apply(state) - valueGetter.apply(s);
                    if (difference > 0) {
                        consumer.accept(
                            time,
                            difference,
                            stack -> List.of(s.apply(stack)),
                            stack -> List.of(state.apply(stack))
                        );
                    }
                });
            } else if (attribute == FERMENTATION) {
                states.stream().filter(s -> !s.vinegar()).forEach(s -> {
                    consumer.accept(
                        time,
                        (FERMENTATION_STEPS + 1) - valueGetter.apply(s),
                        stack -> List.of(s.apply(stack)),
                        stack -> List.of(state.apply(stack))
                    );
                });
            }
        }
    }

    @Override
    public int getHash(ItemStack stack) {
        return Objects.hash(this, settings.variants.find(stack));
    }

    @Override
    public Component getName(ItemStack stack) {
        return settings.variants.find(stack).getName(Component.translatable(getTranslationKey()));
    }

    @Override
    public void appendTooltip(ItemStack stack, @Nullable Level world, List<Component> tooltip, Item.TooltipContext context) {

        int distillation = DISTILLATION.get(stack);
        int maturation = MATURATION.get(stack);
        int fermentation = FERMENTATION.get(stack);

        if (distillation > 0) {
            tooltip.add(Component.translatable("orangesunshine.alcohol.distillations", distillation).withStyle(ChatFormatting.GRAY));
        }

        if (fermentation > 0) {
            tooltip.add(Component.translatable("orangesunshine.alcohol.fermentations", fermentation).withStyle(ChatFormatting.GRAY));
        }

        if (maturation > 0) {
            tooltip.add(Component.translatable("orangesunshine.alcohol.maturations", maturation, Maturity.getMaturity(maturation).getName()).withStyle(ChatFormatting.GRAY));
        }

        //if (distillation > 0 || maturation > 0 || fermentation > 0) {
        //    tooltip.add(Component.empty());
        //    tooltip.add(settings.profile.getFlavour(distillation, fermentation, maturation));
        //}
    }

    @Override
    public int getColor(ItemStack stack) {
        return MathUtils.mixColors(
                MathUtils.mixColors(
                        super.getColor(stack),
                        getDistilledColor(),
                        MathUtils.progress(DISTILLATION.get(stack))
                ),
                getMatureColor(),
                MathUtils.progress(MATURATION.get(stack) * 0.2F)
        );
    }

    @Override
    public Identifier getSymbol(ItemStack stack) {
        return settings.variants.find(stack).getSymbol(getId());
    }

    @Override
    public void getDefaultStacks(FluidContainer container, Consumer<ItemStack> consumer) {
        settings.states.get().forEach(state -> consumer.accept(state.apply(getDefaultStack(container))));
    }

    @SuppressWarnings("deprecation")
    @Override
    public boolean isSuitableContainer(FluidContainer container) {
        return container.asItem().builtInRegistryHolder().is(PSTags.Items.SUITABLE_ALCOHOLIC_DRINK_RECEPTICALS);
    }

    public static class Settings extends DrugFluid.Settings {
        protected DrinkTypes variants = DrinkTypes.empty();

        private double fermentationAlcohol;
        private double distillationAlcohol;
        private double maturationAlcohol;

        private int matureColor = 0xcc592518;
        private int distilledColor = 0x33ffffff;

        DrugType drugType = DrugType.ALCOHOL;

        final Supplier<List<DrinkTypes.State>> states = Suppliers.memoize(() -> variants.streamStates().toList());

        public Supplier<PSConfig.Balancing.FluidProperties.TickInfo> tickInfo;

        public Settings() {
            this.appearance = stack -> variants.find(stack).appearance();
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings drug(DrugType drug) {
            this.drugType = drug;
            return this;
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings matureColor(int matureColor) {
            this.matureColor = matureColor;
            return this;
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings distilledColor(int distilledColor) {
            this.distilledColor = distilledColor;
            return this;
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings variants(DrinkTypes variants) {
            this.variants = variants;
            return this;
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings alcohol(double fermentationAlcohol, double distillationAlcohol, double maturationAlcohol) {
            this.fermentationAlcohol = fermentationAlcohol;
            this.distillationAlcohol = distillationAlcohol;
            this.maturationAlcohol = maturationAlcohol;
            return this;
        }

        public moriz.orangesunshine.fluid.AlcoholicFluid.Settings tickRate(Supplier<PSConfig.Balancing.FluidProperties.TickInfo> tickInfo) {
            this.tickInfo = tickInfo;
            return this;
        }
    }
}
